### ISSUE TITLE
Pyright unable to handle inspect.iscoroutine

### ISSUE THREAD

**Issue by kumaraditya303** [microsoft/pyright#2340]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**

Pyright is unable to narrow the type based on `inspect.iscoroutine`.

**To Reproduce**

```py
import asyncio
import inspect
from typing import Literal


def func(a: bool):
    async def hello() -> Literal["Hello"]:
        await asyncio.sleep(1)
        return "Hello"

    if a:
        return hello()
    else:
        return "Hello"


async def main():
    result = func(False)
    if inspect.iscoroutine(result):
        print(await result)
    else:
        print(result)


asyncio.run(main())


```

**Expected behavior**
There should be no error 

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

```
pyright 1.1.126
```

**Comment by erictraut** [comment]
This isn't a supported type guard. For a full list of supported type guard patterns, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-guards).

Each new type guard requires special-case handling and adds complexity to the type checker and slows down analysis, so we add support for new patterns only if we have a strong signal that it's a frequently used pattern. This doesn't strike me as a common pattern. Mypy doesn't support this pattern either, and it doesn't appear that anyone has requested this feature in the history of mypy's issue tracker.

We can revisit this in the future if there is sufficient interest (as evidenced by upvotes on this issue), but for now I'm going to close this request.

You can work around this using a `cast` or a user-defined TypeGuard (see [PEP 647](https://www.python.org/dev/peps/pep-0647/)). Here's what that would look like:

```pyright
def is_coroutine(value: object) -> TypeGuard[Coroutine[Any, Any, Any]]:
    return inspect.iscoroutine(value)
```

**Comment by kumaraditya303** [comment]
@erictraut Can pyright handle isinstance checks from `types` module ?

I am trying this:

```py
import asyncio
import types
from typing import Literal

def func(a: bool):
    async def hello() -> Literal["Hello"]:
        await asyncio.sleep(1)
        return "Hello"

    if a:
        return hello()
    else:
        return "Hello"


async def main():
    result = func(False)
    if isinstance(result, types.CoroutineType):
        print(await result)
    else:
        print(result)


asyncio.run(main())

```

**Comment by kumaraditya303** [comment]
In typeshed it is defined as a typeguard as you suggested see https://github.com/python/typeshed/blob/a9227ed98589bf960b1c9ee2da69890171f8dfd1/stdlib/inspect.pyi#L65

**Comment by erictraut** [comment]
No, there is no association between the type `typing.Coroutine` and `types.CoroutineType`, and pyright doesn't contain any special-casing for this association.

**Comment by kumaraditya303** [comment]
@erictraut Apparently it was a typeshed bug see https://github.com/python/typeshed/issues/6086 hence I was getting this error 

```console
D:\playwright-python\playwright\_impl\_browser_context.py:156:41 - error: Argument of type "CoroutineType" cannot be assigned to parameter "coro" of type "Generator[_TaskYieldType, None, _T@create_task] | Awaitable[_T@create_task]" in function "create_task"
    Type "CoroutineType" cannot be assigned to type "Generator[_TaskYieldType, None, _T@create_task] | Awaitable[_T@create_task]"
      "CoroutineType" is incompatible with "Generator[_TaskYieldType, None, _T@create_task]"
      "CoroutineType" is incompatible with protocol "Awaitable[_T@create_task]"
        "__await__" is not present (reportGeneralTypeIssues)
``` 
Can you reopen this issue until you roll pyright's typeshed copy to the fixed one ?

**Comment by erictraut** [comment]
Ah, good fix!

**Comment by kumaraditya303** [comment]
This is the commit by me which fixed it https://github.com/python/typeshed/commit/25ae5b887194a0b6178bc7eeeb0ab6db6a4dcd4c

**Comment by erictraut** [comment]
I've pulled the latest typeshed stubs into the pyright repo, so this will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.173, which I just published. It will also be included in the next release of pylance.
